src/puya/awst_build/eb/arc4/abi_call.py

```diff
@@ -154,40 +186,174 @@ def call(
         )
 
 
+def _get_python_kwargs(fields: Sequence[TxnField]) -> Set[str]:
+    return StableSet.from_iter(
+        arg for arg, param in PYTHON_ITXN_ARGUMENTS.items() if param.field in fields
+    )
+
+
+class _ARC4CompilationFunctionBuilder(FunctionBuilder):
+    allowed_fields: Sequence[TxnField]
+
+    @property
+    def allowed_kwargs(self) -> Set[str]:
+        return {_COMPILED_KWARG, *_get_python_kwargs(self.allowed_fields)}
+
+    @typing.override
+    def call(
+        self,
+        args: Sequence[NodeBuilder],
+        arg_kinds: list[mypy.nodes.ArgKind],
+        arg_names: list[str | None],
+        location: SourceLocation,
+    ) -> InstanceBuilder:
+        # if on_completion not allowed, it must be an update
+        is_update = TxnField.OnCompletion not in self.allowed_fields
+        method_or_type, abi_args, kwargs = _get_method_abi_args_and_kwargs(
+            args, arg_names, self.allowed_kwargs
+        )
+        compiled: InstanceBuilder | None = None
+        if compiled_node := kwargs.pop(_COMPILED_KWARG, None):
+            compiled = expect.exactly_one_arg_of_type_else_dummy(
+                [compiled_node], pytypes.CompiledContractType, compiled_node.source_location
+            )
+
+        match method_or_type:
+            case None:
+                raise CodeError("missing required positional argument 'method'", location)
+            case BaseClassSubroutineInvokerExpressionBuilder(
+                target=BaseClassSubroutineTarget(base_class=base)
+            ) as eb:
+                method_call = _get_arc4_method_call(eb.context, eb.node, abi_args, location)
+                if compiled is None:
+                    compiled = CompiledContractExpressionBuilder(
+                        CompiledContract(
+                            contract=base,
+                            wtype=pytypes.CompiledContractType.wtype,
+                            source_location=location,
+                        )
+                    )
+                else:
+                    _warn_if_different_contract(
+                        compiled, base, related_location=eb.source_location
+                    )
+            case ContractTypeExpressionBuilder(
+                pytype=pytypes.TypeType(typ=typ),
+                type_info=type_info,
+            ) as eb if pytypes.ARC4ContractBaseType in typ.mro:
+                method_call = _get_lifecycle_method_call(
+                    eb.context,
+                    type_info,
+                    abi_args,
+                    kind="update" if is_update else "create",
+                    location=method_or_type.source_location,
+                )
+                module_name, class_name = type_info.fullname.rsplit(".", maxsplit=1)
+                contract_ref = ContractReference(
+                    module_name=module_name,
+                    class_name=class_name,
+                )
+                if compiled is None:
```

Comment:
this if-else is essentially the same as the on in the case above - can it be moved outside simply?